libyul/optimiser/KnowledgeBase.cpp

```diff
@@ -75,34 +66,144 @@ bool KnowledgeBase::knownToBeDifferentByAtLeast32(YulString _a, YulString _b)
 {
-	if (m_counter++ > 100)
-		return _expression;
+	if (Literal const* literal = std::get_if<Literal>(&_value))
+		return VariableOffset{YulString{}, valueOfLiteral(*literal)};
+	else if (Identifier const* identifier = std::get_if<Identifier>(&_value))
+		return explore(identifier->name);
+	else if (FunctionCall const* f = get_if<FunctionCall>(&_value))
+	{
+		if (f->functionName.name == "add"_yulstring)
+		{
+			if (optional<VariableOffset> a = explore(f->arguments[0]))
+				if (optional<VariableOffset> b = explore(f->arguments[1]))
+				{
+					u256 offset = a->offset + b->offset;
+					if (a->reference.empty())
+						// a is constant
+						return VariableOffset{b->reference, offset};
+					else if (b->reference.empty())
+						// b is constant
+						return VariableOffset{a->reference, offset};
+				}
+		}
+		else if (f->functionName.name == "sub"_yulstring)
+			if (optional<VariableOffset> a = explore(f->arguments[0]))
+				if (optional<VariableOffset> b = explore(f->arguments[1]))
+				{
+					u256 offset = a->offset - b->offset;
+					if (a->reference == b->reference)
+						return VariableOffset{YulString{}, offset};
+					else if (b->reference.empty())
+						// b is constant
+						return VariableOffset{a->reference, offset};
+				}
+	}
+
+	return nullopt;
+}
 
-	if (holds_alternative<FunctionCall>(_expression))
-		for (Expression& arg: std::get<FunctionCall>(_expression).arguments)
-			arg = simplifyRecursively(arg);
+Expression const* KnowledgeBase::valueOf(YulString _var)
+{
+	AssignedValue const* assignedValue = m_variableValues(_var);
+	Expression const* currentValue = assignedValue ? assignedValue->value : nullptr;
+	if (m_valuesAreSSA)
+		return currentValue;
+
+	Expression const* lastValue = m_lastKnownValue[_var];
+	if (lastValue != currentValue)
+		reset(_var);
+	m_lastKnownValue[_var] = currentValue;
+	return currentValue;
+}
 
-	if (auto match = SimplificationRules::findFirstMatch(_expression, m_dialect, m_variableValues))
-		return simplifyRecursively(match->action().toExpression(debugDataOf(_expression)));
+void KnowledgeBase::reset(YulString _var)
+{
+	yulAssert(!m_valuesAreSSA);
+
+	m_lastKnownValue.erase(_var);
+	if (VariableOffset const* offset = util::valueOrNullptr(m_offsets, _var))
+	{
+		// Remove var from its group
+		if (!offset->reference.empty())
+			m_groupMembers[offset->reference].erase(_var);
+		m_offsets.erase(_var);
+	}
+	if (set<YulString>* group = util::valueOrNullptr(m_groupMembers, _var))
+	{
+		// _var was a representative, we might have to find a new one.
+		if (group->empty())
+			m_groupMembers.erase(_var);
+		else
+		{
+			YulString newRepresentative = *group->begin();
+			u256 newOffset = m_offsets[newRepresentative].offset;
+			for (YulString groupMember: *group)
+			{
+				yulAssert(m_offsets[groupMember].reference == _var);
+				m_offsets[groupMember].reference = newRepresentative;
+				m_offsets[newRepresentative].offset -= newOffset;
```

Comment:
Oh, you are right, we also need `m_groupMembers[newRepresentative] = std::move(m_groupMembers[_var])`.

I'll add a comment about why we subtract here.

Reply:
Oh wow, this should be `m_offsets[groupMember]` shouldn't it?

Reply:
I assumed that every `groupMember` would have the same offset, but now I see it is only the same reference, right?
So, I think `m_offsets[groupMember]` should be updated. But then, shouldn't it be just an assignment instead of decrease?
`m_offsets[groupMember].offset = newOffset;` 
instead of 
`m_offsets[groupMember].offset -= newOffset;`
?